r3bbase/pars/R3BMSOffsetFinder.cxx

```diff
@@ -0,0 +1,179 @@
+    : R3BMSOffsetFinder("R3B MS Offset Finder", 1)
+{
+}
+
+// R3BMSOffsetFinder: Standard Constructor --------------------------
+R3BMSOffsetFinder::R3BMSOffsetFinder(const TString& name, Int_t iVerbose)
+    : FairTask(name, iVerbose)
+    , fMSOffsetPar(NULL)
+    , fSamplerMapped(NULL)
+    , fSamplerMSMapped(NULL)
+    , fMinStadistics(1000)
+{
+}
+
+// R3BMSOffsetFinder: Destructor ----------------------------------------
+R3BMSOffsetFinder::~R3BMSOffsetFinder() { LOG(info) << "R3BMSOffsetFinder: Delete instance"; }
+
+void R3BMSOffsetFinder::SetParContainers()
+{
+    // Parameter Container
+    // Reading amsMappingPar from FairRuntimeDb
+    FairRuntimeDb* rtdb = FairRuntimeDb::instance();
+    if (!rtdb)
+    {
+        LOG(error) << "FairRuntimeDb not opened!";
+    }
+}
+
+// -----   Public method Init   --------------------------------------------
+InitStatus R3BMSOffsetFinder::Init()
+{
+
+    LOG(info) << "R3BMSOffsetFinder: Init";
+
+    FairRootManager* rootManager = FairRootManager::Instance();
+    if (!rootManager)
+    {
+        return kFATAL;
+    }
+
+    fSamplerMapped = dynamic_cast<TClonesArray*>(rootManager->GetObject("SamplerMapped"));
+    if (!fSamplerMapped)
+    {
+        return kFATAL;
+    }
+
+    fSamplerMSMapped = dynamic_cast<TClonesArray*>(rootManager->GetObject("SamplerMSMapped"));
+    if (!fSamplerMSMapped)
+    {
+        return kFATAL;
+    }
+
+    FairRuntimeDb* rtdb = FairRuntimeDb::instance();
+    if (!rtdb)
+    {
+        return kFATAL;
+    }
+
+    fMSOffsetPar = dynamic_cast<R3BMSOffsetPar*>(rtdb->getContainer("MSOffsetPar"));
+    if (!fMSOffsetPar)
+    {
+        LOG(error) << "R3BMSOffsetFinder::Init() Couldn't get handle on MSOffsetPar container";
+        return kFATAL;
+    }
+
+    fh_Offset_Finder = new TH1F("MSOffsetFinderHisto", "MSOffsetFinderHisto", 2000, -1000, 1000);
+
+    return kSUCCESS;
+}
+
+// -----   Public method ReInit   --------------------------------------------
+InitStatus R3BMSOffsetFinder::ReInit()
+{
+    SetParContainers();
+    return kSUCCESS;
+}
+
+// -----   Public method Exec   --------------------------------------------
+void R3BMSOffsetFinder::Exec(Option_t* opt)
+{
+
+    Int_t sampHits = fSamplerMapped->GetEntriesFast();
+    Int_t sampmsHits = fSamplerMSMapped->GetEntriesFast();
+    if (!sampHits && !sampmsHits)
+        return;
+    Double_t SAMPTime = 0;
+    Double_t SAMPMSTime = 0;
+    R3BSamplerMappedData* SAMPMapped;
+    R3BSamplerMappedData* SAMPMSMapped;
+
+    for (Int_t i = 0; i < sampmsHits; i++)
+    {
+        SAMPMSMapped = dynamic_cast<R3BSamplerMappedData*>(fSamplerMSMapped->At(i));
+        SAMPMSTime = SAMPMSMapped->GetTime();
+    }
+
+    for (Int_t i = 0; i < sampHits; i++)
+    {
+        SAMPMapped = dynamic_cast<R3BSamplerMappedData*>(fSamplerMapped->At(i));
+        SAMPTime = SAMPMapped->GetTime();
+
+        // Fill the histogram
+        fh_Offset_Finder->Fill(SAMPTime - SAMPMSTime);
+    }
```

Comment:
Not sure this is correct. `SAMPMSTime` value is overwritten in the loop and only the last entry is saved.

Reply:
It is supposed to have only one value for SAMPMSTime in each event. If it doesn't occur this is a problem. I make the loop because it can happen to have more than one value here, this way writing only the last one will be enough for the next step that will be implemented in the near future (The calculations for Tprev and Tnext).

Reply:
I changed the conditions in the previous if statement, so now I eliminated this "for loop".